messages/jsonschema/Pickle.json

```diff
@@ -42,6 +44,21 @@
           "description": "A unique ID for the PickleStep",
           "type": "string"
         },
+        "keyword": {
+          "description": "The keyword as it appears in the feature file",
+          "type": "string"
+        },
+        "keywordType": {
+          "description": "The context in which the step was specified: context (Given), action (When) or outcome (Then).\n\nNote that the keywords `But` and `And` inherit their meaning from prior steps and the `*` 'keyword' doesn't have specific meaning (hence General)",
+          "type": "string",
+          "enum": [
```

Comment:
I think the following enum values here are sufficient:

* `Context`
* `Action`
* `Outcome`
* `Unknown`

At the pickle level, it should make no difference wherher an outcome step had the keyword `Then`, `And` or `But` in the source. The pickle compiler should translate `And` and `But` to one of `Context`, `Action`, `Outcome` (or `Unknown` in the unlikely event that the preceding source keyword was a `* `.

Not having `Conjunction` at the pickle step level also makes interpretation easier, since the interpretation can be encapsulated inside the compiler so that pickle consumers don't need to do it.

Reply:
When the pickle is focused on execution, I think you have a strong point. I also started out like that (although the names of the keywords were different). However, @mpkorstanje and @sebrose(?) suggested in [prior discussion](https://github.com/cucumber/common/pull/1741#issuecomment-916129760) that by compiling conjunctions into the keywords they refer to, we loose information. Between the Pickle and the GherkinDocument, we don't, their point is that the scenario can't be rendered from the Pickle alone.

Reply:
Pickles are only meant to be used for execution - not reporting. By delegating the responsibility to the pickle consumer to figure out whether a step is context/action/outcome, we're opening up for multiple inconsistent interpretations for no obvious benefit.

We do lose information, but it's information we don't need for execution. The purpose of this PR (as I see it) is simply to be able to prevent the use of a `Given` step definition from a `When` step etc, and I think we should make this as easy as possible to implement on the Cucumber side.

Reply:
> Pickles are only meant to be used for execution - not reporting. By delegating the responsibility to the pickle consumer to figure out whether a step is context/action/outcome, we're opening up for multiple inconsistent interpretations for no obvious benefit.

Fundamentally this isn't true. The feature format is not an accurate representation of the what is executed.

When reporting on what will be excuted a reasonable mapping to the feature file is needed. This includes the keyword as it occurs in the feature file.

Reply:
@mpkorstanje Isn't it the case that @aslakhellesoy tries to separate the "reporting side" (including the "this is what I'm executing" output) from the actual execution ("please execute this pickle step with this keyword type, this argument and these parameters")?

I do see why that then later becomes a challenge to correctly report to the user *if* you're doing that based on the pickle content. I'm thinking that @aslakhellesoy would want the progress report to be based on the "testCase" and the AST, so that you don't need the pickle. (The AST has the keyword, keyword type, argument and parameters, all of which you need to do the reporting; the testCase has the test outcome, which would help render in the correct color?)